readme.md
-------

# SCM MUSIC PLAYER FOR SMF v2.5

[**By Dougiefresh**](http://www.simplemachines.org/community/index.php?action=profile;u=253913) -> [Link to Mod](http://custom.simplemachines.org/mods/index.php?mod=4132)

-------

## Introduction
This mod adds the ability to start playing MP3, WAV or OGG file(s) in the background automatically using the [SMC Media Player](http://scmplayer.net/).

**## Admin Settings**
Under **Admin** => **Configuration** => **Modification Settings**, there is a new section called **SCM Music Player**.  In that area, many settings of this mod can be controlled, such as:

- **Enable playing background music?**
- **Background music volume level**
- **Autoplay the background music?**
- **Shuffle music playlist before starting?**
- **Repeat mode:**
- **Placement of Player Bar**
- **Show playlist by default? **
- SMC media player skin selection.

With version 1.2, you can disable the player in the following areas:

- On Board Index
- On Message Index
- In Topics
- In My Messages
- In Profile
- In Admin area
- In Moderate area
- In Calendar

There is also a new permission under **Admin** => **Forum** => **Permissions**:

- **Listen to Site Music**

## Related Discussion
- [Music Mod?](http://www.simplemachines.org/community/index.php?topic=550549.0)

## Compatibility Notes
This mod was tested on SMF 2.0.12, but should work on any version of SMF 2.0.x.  SMF 1.x and SMF 2.1 is not and will not be supported.

## Translators

- Spanish Latin: [Rock Lee](https://www.simplemachines.org/community/index.php?action=profile;u=322597)

## Changelog
The changelog can be viewed at [XPtsp.com](http://www.xptsp.com/board/free-modifications/scm-music-player-for-smf/?tab=1).

## License
Copyright (c) 2016 - 2018, Douglas Orend

All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changelog.txt
v2.5 - October 23rd, 2018
o No functionality change.
o Updated documentation to point to new website.

v2.4 - May 21st, 2018
o Updated Spanish Latin translation, courtsey of [url=https://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee[/url].
o Fixed [b]readme.txt[/b] to [b]Spanish Latin[/b] from [b]Spanish Lation[/b].

v2.3 - April 19th, 2018
o Fixed undefined index error in [b]Subs-SCMP.php[/b] by loading language file.
o [b]SMF 2.1[/b]: Fixed undefined index error in [b]SCMP.template.php[/b].

v2.2 - March 24th, 2018
o Fixed inability to add new playlists.
o Fixed code that saves playlists to the database.
o Fixed undefined variable error in [b]SCMP.template.php[/b].
o Removed references to missing CSS file.
o Moved playlist buidling code from [b]Subs-SCMP_Admin.php[/b] to [b]Subs-SCMP.php[/b].
o Added code in order to cache the building of the playlists.

v2.1 - March 15th, 2018
o Fixed issue involving selecting playlists other than number 0 (the default).

v2.0 - February 28th, 2018
o Added ability to add multiple playlists.
o Admin can choose which playlist gets played by every user.

v1.10 - February 12th, 2018
o Added Spanish Latin translation, courtsey of [url=https://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee[/url].
o Added translators section to [b]readme.txt[/b] documentation.
o Added [b]LICENSE[/b] file, to clarify licensing terms.

v1.9 - January 7th, 2018
o Added time parameters for restricting autoplay setting.

v1.8 - August 19th, 2017
o Changed GitHub source from [b]cshum/scm-music-player[/b] to [b]icze4r/scm-music-player-SSL-[/b]

v1.7 - August 18th, 2017
o Fixed issue where enabling the player apparently didn't take.
o Added source files to SCM Music Player to mod.
o Mod now pulls SCM Music Player from the forum folder.

v1.6 - August 10th, 2017
o Fixed a lot of decision-making code to make the SCM player work properly.

v1.5 - August 9th, 2017
o Removed debug code from main admin hook.

v1.4 - July 5th, 2017
o Fixed the hook that adds new section to the admin panel, which caused issues in Core Features.

v1.3 - March 27th, 2017
o Added check to see if function safe_serialize exists before use.

v1.2 - March 23th, 2017
o Changed admin area for mod into its own seperate admin area.
o Admin area now has 3 subsections: Settings, Playlists, Skins.
o Added various options to disable player for multiple areas.

v1.1 - February 22nd, 2017
o Added "Listen To Site Music" permission to the mod.

v1.0.2 - January 7th, 2017
o Replaced invalid redirection URL in [b]package-info.xml[/b].
o Replaced deprecated [b]center[/b] tag with CSS directives to center the table.
o Added code to make sure that a skin style was defined if the variable wasn't set.
o Changed code so that empty title/url combo was added after playlist.

v1.0.1 - December 19th, 2016
o Complete rewrite, using SCM Music Player as basis instead of pure HTML5 audio tag.

v1.0 - November 2nd, 2016
o Initial Release

Subs-SCMP.php
<?php
/********************************************************************************
* Subs-SCMP.php - Subs of the SCM Music Player for SMF mod
*********************************************************************************
* This program is distributed in the hope that it is and will be useful, but
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY
* or FITNESS FOR A PARTICULAR PURPOSE,
**********************************************************************************/
if (!defined('SMF'))
	die('Hacking attempt...');

/*******************************************************************************/
// Hook functions:
/*******************************************************************************/
function SCM_Load()
{
	global $context, $modSettings, $txt, $boardurl, $user_info, $board, $topic;

	// Disable the player if we are not allowed to listen to Site Music:
	$disabled = !allowedTo('listen_to_music');
	$action = isset($_GET['action']) && $_GET['action'] != 'forum' ? $_GET['action'] : false;
	$enabled = $modSettings['SCM_enabled'];
	if (!$disabled)
	{
		$disabled |= (!empty($modSettings['SCM_hide_boardindex']) && empty($action) && empty($board) && empty($topic));
		$disabled |= (!empty($modSettings['SCM_hide_messageindex']) && empty($action) && !empty($board) && empty($topic));
		$disabled |= (!empty($modSettings['SCM_hide_posts']) && !empty($topic));
		$disabled |= (!empty($modSettings['SCM_hide_profile']) && $action == 'profile');
		$disabled |= (!empty($modSettings['SCM_hide_pm']) && $action == 'pm');
		$disabled |= (!empty($modSettings['SCM_hide_members']) && $action == 'members');
		$disabled |= (!empty($modSettings['SCM_hide_admin']) && $action == 'admin');
		$disabled |= (!empty($modSettings['SCM_hide_calendar']) && $action == 'calendar');
		$disabled |= (!empty($modSettings['SCM_hide_moderate']) && $action == 'moderate');
	}
	if ($disabled)
		$_SESSION['SCM_last_update'] = $enabled = false;

	// Build the SCMP playlist array if not cached:
	if (($context['SCMP_playlists'] = cache_get_data('SCMP_playlists', 86400)) == null)
	{
		// Load the language file we need for this area:
		loadLanguage('SCMP');

		// Gather everything we need for the "default" playlist:
		$func = function_exists('safe_unserialize') ? 'safe_unserialize' : 'unserialize';
		$songs = !empty($modSettings['SCM_playlist']) ? $modSettings['SCM_playlist'] : '';
		$songs = @$func($songs);
		if (empty($songs) || !is_array($songs))
			$songs = array();
		$context['SCMP_playlists'] = array(
			0 => array(
				'id' => 0,
				'name' => !empty($songs['__NAME__']) ? $songs['__NAME__'] : $txt['SCM_playlists_Default'],
			),
		);
		unset($songs['__NAME__']);
		$context['SCMP_playlists'][0]['songs'] = $songs;	

		// Gather up any other playlists available in the $modSettings array:
		foreach ($modSettings as $variable => $value)
		{
			if (preg_match('~SCM_playlist_([\d]+)~i', $variable, $matches) && !empty($value))
			{
				$songs = @$func($value);
				if (empty($songs) || !is_array($songs))
					$songs = array();
				$id = $matches[1];
				$context['SCMP_playlists'][$id] = array(
					'id' => $id,
					'name' => !empty($songs['__NAME__']) ? $songs['__NAME__'] : sprintf($txt['SCMP_playlist_which'], $id),
				);
				unset($songs['__NAME__']);
				$context['SCMP_playlists'][$id]['songs'] = $songs;
			}
		}
		asort($context['SCMP_playlists']);
		
		// Save the playlist information in the SMF cache:
		cache_put_data('SCMP_playlists', $context['SCMP_playlists'], 86400);
	}

	// Does the specified playlist exist?  If not, return to caller:
	$selected = !empty($modSettings['SCM_selected_playlist']) ? (int) $modSettings['SCM_selected_playlist'] : 0;
	if (empty($context['SCMP_playlists'][$selected]['songs']))
		return;
		
	// Should we revise the playlist currently playlist?
	$_SESSION['SCMP_selected'] = $selected;

	// Do we need to update the current player?  If not, return to caller:
	if (!empty($_SESSION['SCM_last_update']) && !empty($modSettings['SCM_last_update']) && $_SESSION['SCM_last_update'] < $modSettings['SCM_last_update'])
		return;
	$_SESSION['SCM_last_update'] = time();

	// Build the playlist for the Javascript code:
	$echo = array();
	foreach ($context['SCMP_playlists'][$selected]['songs'] as $title => $url)
		$echo[$title] = '{\'title\':' . JavaScriptEscape($title) . ',\'url\':' . JavaScriptEscape($url) . '}';
	$playlist = implode($echo, ',');

	// Decide on the CSS for the player:
	if (!$enabled || empty($playlist))
		$css = $boardurl . '/Themes/default/css/hide_player.css';
	elseif (!empty($modSettings['SCM_style']))
	{
		if ($modSettings['SCM_style'] == 'custom' && !empty($modSettings['SCM_custom_url']))
			$css = $modSettings['SCM_custom_url'];
		else
			$css = 'skins/' . (!isset($modSettings['SCM_style']) ? 'aquaBlue' : $modSettings['SCM_style']) . '/skin.css';
	}
	if (empty($css))
		$css = 'skins/aquaBlue/skin.css';
	if (substr($css, 0, 4) !== 'http')
		$css = $boardurl . '/SCM_Music_Player/' . $css;

	// Save some decisions that need to be made in the code:
	$placement = !empty($modSettings['SCM_placement']) && $modSettings['SCM_placement'] == 'bottom' ? 'bottom' : 'top';
	$autoplay = !empty($modSettings['SCM_autoplay']);
	if ($autoplay && !empty($modSettings['SCM_autoplay_from']) && !empty($modSettings['SCM_autoplay_to']))
	{
		$hour = (int) date('G');
		if ($modSettings['SCM_autoplay_from'] > $modSettings['SCM_autoplay_to'])
			$autoplay = ($hour >= $modSettings['SCM_autoplay_from']) || ($hour <= $modSettings['SCM_autoplay_to']);
		else
			$autoplay = ($hour >= $modSettings['SCM_autoplay_from'] && $hour <= $modSettings['SCM_autoplay_to']);
	}

	// Insert the player into the HTML header:
	$context['html_headers'] .= '
	<!-- SCM Music Player http://scmplayer.net -->
	<script type="text/javascript" src="' . $boardurl . '/SCM_Music_Player/script.js" data-config="{
		\'skin\': \'' . $css . '\',
		\'volume\': ' . ((int) (!isset($modSettings['SCM_volume']) ? 50 : $modSettings['SCM_volume'])) . ',
		\'autoplay\': ' . ($enabled && $autoplay ? 'true' : 'false'). ',
		\'shuffle\': ' . (!empty($modSettings['SCM_shuffle']) ? 'true' : 'false') . ',
		\'repeat\': ' . (!isset($modSettings['SCM_repeat']) ? '1' : $modSettings['SCM_repeat']) . ',
		\'placement\': \'' . $placement . '\',
		\'showplaylist\': ' . (!empty($modSettings['SCM_show_playlist']) ? 'true' : 'false') . ',
		\'playlist\': [' . $playlist . ']}">
	</script>';

	// Let's make any changes that the admin have made to the player:
	$context['html_headers'] .= '
	<script type="text/javascript">';

	// Did the placement of the SCMP player change?
	if (isset($_SESSION['SCMP_placement']) && $_SESSION['SCMP_placement'] <> $placement)
		$context['html_headers'] .= '
		SCM.placement("' . $placement . '");';
	$_SESSION['SCMP_placement'] = $placement;

	// Did the selected/current playlist change?
	if (isset($_SESSION['SCMP_selected']) && $_SESSION['SCMP_selected'] <> $selected)
		$context['html_headers'] .= '
		SCM.loadPlaylist([' . $playlist . ']);';
	$_SESSION['SCMP_selected'] = $selected;

	// Is the SCMP player disabled?
	$context['html_headers'] .= '
		if (SCM.isPlay()) {
			SCM.' . ($enabled ? 'start' : 'stop') . '();
		}';

	// Finish the header stuff for this mod:
	$context['html_headers'] .= '
	</script>
	<!-- SCM Music Player script end -->';
}

?>

SCMP.english.php
<?php
/********************************************************************************
* SiteMusic.english.php - English language file for Site Music mod
*********************************************************************************
* This program is distributed in the hope that it is and will be useful, but
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY
* or FITNESS FOR A PARTICULAR PURPOSE,
**********************************************************************************/

$txt['permissionname_listen_to_music'] = 'Can Listen to Site Music';

$txt['SCM_area'] = 'SCM Music Player';
$txt['SCM_description'] = '';
$txt['SCM_title'] = 'SCM Music Player settings';

//$txt['language_settings'] = 'Settings';
$txt['SCM_enabled'] = 'Enable playing background music?';
$txt['SCM_style'] = 'Player Style:';
$txt['SCM_autoplay'] = 'Autoplay the background music?';
$txt['SCM_autoplay_from'] = 'Start autoplay at what hour?<div class="smalltext"><strong>NOTE:</strong> 1-24 or set to <strong>0</strong> to play all day</div>';
$txt['SCM_autoplay_to'] = 'Stop autoplay at end of what hour?<div class="smalltext"><strong>NOTE:</strong> 1-24 or set to <strong>0</strong> to play all day</div>';
$txt['SCM_volume'] = 'Background music volume level<div class="smalltext"><strong>NOTE:</strong> Must be between <strong>0</strong> and <strong>100</strong></div>';
$txt['SCM_shuffle'] = 'Shuffle music playlist before starting?';
$txt['SCM_repeat'] = 'Repeat mode:';
$txt['SCM_repeat0'] = 'Play Playlist Once';
$txt['SCM_repeat1'] = 'Repeat Playlist';
$txt['SCM_repeat2'] = 'Repeat Item';
$txt['SCM_placement'] = 'Placement of Player Bar';
$txt['SCM_top'] = 'Top';
$txt['SCM_bottom'] = 'Bottom';
$txt['listen_to_music'] = 'Membergroups can listen to Site Music:';
$txt['SCM_show_playlist'] = 'Show playlist by default?';
$txt['SCM_hide_boardindex'] = 'Hide Blocks in Forum Areas';
$txt['SCM_boardindex'] = 'Board Index';

$txt['SCM_playlist'] = 'Playlist';
$txt['SCM_playlist_title'] = 'List of Tracks';
$txt['SCM_music_title'] = 'Song Title';
$txt['SCM_music_URL'] = 'Song URL';
$txt['SCM_add_track'] = 'Add Track';

$txt['SCM_skins'] = 'Skins';
$txt['SCM_style_title'] = 'Style &amp; Color Options';
$txt['SCM_custom_style'] = 'Custom URL:';

$txt['SCM_playlists'] = 'Playlists';
$txt['SCM_playlists_active'] = 'Active';
$txt['SCM_playlists_name'] = 'Name of Playlist';
$txt['SCM_playlists_Default'] = 'Default playlist';
$txt['SCM_edit_playlist'] = 'Song List';
$txt['SCM_new_playlist'] = 'New Playlist';
$txt['SCMP_playlist_which'] = 'Playlist %d';
$txt['SCMP_remove'] = 'remove';

$txt['SCM_playlist_nonexistant'] = 'The specified playlist does not exist!';
$txt['SCM_cannot_remove_default'] = 'Cannot remove the default playlist!';

?>

add_remove_hooks.php
<?php
/**********************************************************************************
* add_remove_hooks.php                                                            *
***********************************************************************************
***********************************************************************************
* This program is distributed in the hope that it is and will be useful, but      *
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY    *
* or FITNESS FOR A PARTICULAR PURPOSE.                                            *
*                                                                                 *
* This file is a simplified database installer. It does what it is suppoed to.    *
**********************************************************************************/

// If we have found SSI.php and we are outside of SMF, then we are running standalone.
if (file_exists(dirname(__FILE__) . '/SSI.php') && !defined('SMF'))
	require_once(dirname(__FILE__) . '/SSI.php');
elseif (!defined('SMF')) // If we are outside SMF and can't find SSI.php, then throw an error
	die('<b>Error:</b> Cannot install - please verify you put this file in the same place as SMF\'s SSI.php.');
if (SMF == 'SSI')
	db_extend('packages');
	
// Define the hooks
$hook_functions = array(
/* SMF 2.x hooks */
	'integrate_pre_include' => '$sourcedir/Subs-SCMP.php',
	'integrate_load_theme' => 'SCM_Load',
/* SMF 2.x Admin hooks */
	'integrate_admin_include' => '$sourcedir/Subs-SCMP_Admin.php',
	'integrate_admin_areas' => 'SCM_Admin',
	'integrate_load_permissions' => 'SCM_Permissions',
);

// Adding or removing them?
if (!empty($context['uninstalling']))
	$call = 'remove_integration_function';
else
	$call = 'add_integration_function';

// Do the deed
foreach ($hook_functions as $hook => $function)
	$call($hook, $function);

if (SMF == 'SSI')
   echo 'Congratulations! You have successfully installed this mod!';

?>

Subs-SCMP_Admin.php
<?php
/********************************************************************************
* Subs-SCMP_Admin.php - Admin subs of the SCM Music Player for SMF mod
*********************************************************************************
* This program is distributed in the hope that it is and will be useful, but
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY
* or FITNESS FOR A PARTICULAR PURPOSE,
**********************************************************************************/
if (!defined('SMF'))
	die('Hacking attempt...');

/*******************************************************************************/
// Hook functions:
/*******************************************************************************/
function SCM_Admin(&$areas)
{
	global $txt, $context;
	loadLanguage('SCMP');
	$tmp = $context['SCMP_decision'] = isset($_REQUEST['area']) && $_REQUEST['area'] == 'scm_media_player' && isset($_REQUEST['sa']) && $_REQUEST['sa'] == 'edit';
	$areas['config']['areas']['scm_media_player'] = array(
		'label' => $txt['SCM_area'],
		'function' => 'SCM_Config',
		'icon' => 'modifications.gif',
		'subsections' => array(
			'settings' => array($txt['language_settings']),
			'skins' => array($txt['SCM_skins']),
			'playlists' => array($txt['SCM_playlists']),
			($tmp ? 'edit' : 'new') => array($txt[$tmp ? 'SCM_edit_playlist' : 'SCM_new_playlist']),
		),
	);
}

function SCM_Permissions(&$permissionGroups, &$permissionList, &$leftPermissionGroups, &$hiddenPermissions, &$relabelPermissions)
{
	$permissionList['membergroup']['listen_to_music'] = array(false, 'general', 'view_basic_info');
}

/*******************************************************************************/
// Function required for mod configuration:
/*******************************************************************************/
function SCM_Config()
{
	global $context, $settings, $txt, $sourcedir, $modSettings;

	// Create the tabs for the template .
	$context[$context['admin_menu_name']]['tab_data'] = array(
		'title' => $txt['SCM_area'],
		'description' => $txt['SCM_description'],
		'tabs' => array(
			'settings' => array(
			),
			'skins' => array(
			),
			'playlists' => array(
			),
			'edit' => array(
			),
			'new' => array(
			),
		),
	);
	unset($context[$context['admin_menu_name']]['tab_data']['tabs'][$context['SCMP_decision'] ? 'new' : 'edit']);

	// Format: 'sub-action' => 'function'
	$subActions = array(
		'playlists' => 'SCMP_Lists',
		'skins' => 'SCMP_Skins',
		'settings' => 'SCMP_Settings',
		'new' => 'SCMP_Edit',
		'edit' => 'SCMP_Edit',
		'remove' => 'SCMP_Remove',
	);

	// Figure out what we are loading for the user:
	loadTemplate('SCMP');
	require_once($sourcedir . '/ManageServer.php');
	$_REQUEST['sa'] = isset($_REQUEST['sa']) && isset($subActions[$_REQUEST['sa']]) ? $_REQUEST['sa'] : 'settings';
	$context['page_title'] = $txt['SCM_area'];
	return $subActions[$_REQUEST['sa']]();
}

/*******************************************************************************/
// Function handling changing mod settings:
/*******************************************************************************/
function SCMP_Settings($return_config = false)
{
	global $txt, $scripturl, $context, $settings, $modSettings;

	$config_vars = array(
		array('check', 'SCM_enabled'),
		array('permissions', 'listen_to_music'),
		'',
		array('check', 'SCM_autoplay'),
		array('int', 'SCM_autoplay_from'),
		array('int', 'SCM_autoplay_to'),
		'',
		array('int', 'SCM_volume'),
		array('check', 'SCM_shuffle'),
		array('select', 'SCM_repeat', array(0 => $txt['SCM_repeat0'], 1 => $txt['SCM_repeat1'], 2 => $txt['SCM_repeat2'])),
		array('select', 'SCM_placement', array('top' => $txt['SCM_top'], 'bottom' => $txt['SCM_bottom'])),
		array('check', 'SCM_show_playlist'),
		'',
		array('check', 'SCM_hide_boardindex', 'postinput' => $txt['SCM_boardindex']),
		array('check', 'SCM_hide_messageindex', 'postinput' => $txt['message_index']),
		array('check', 'SCM_hide_posts', 'postinput' => $txt['posts']),
		array('check', 'SCM_hide_profile', 'postinput' => $txt['profile']),
		array('check', 'SCM_hide_pm', 'postinput' => $txt['pm_short']),
		array('check', 'SCM_hide_members', 'postinput' => $txt['members']),
		array('check', 'SCM_hide_admin', 'postinput' => $txt['admin']),
		array('check', 'SCM_hide_calendar', 'postinput' => $txt['calendar']),
		array('check', 'SCM_hide_moderate', 'postinput' => $txt['moderate']),
	);
	if ($return_config)
		return $config_vars;

	// Saving?
	if (isset($_GET['save']))
	{
		checkSession();
		$_POST['SCM_last_update'] = time();
		$config_vars[] = array('int', 'SCM_last_update');

		// Save the settings for this mod:
		saveDBSettings($config_vars);
		redirectexit('action=admin;area=scm_media_player');
	}

	// Make sure certain settings are defaulted if not set:
	$modSettings['SCM_style'] = !empty($modSettings['SCM_style']) ? $modSettings['SCM_style'] : 'aquaBlue';
	$modSettings['SCM_volume'] = !isset($modSettings['SCM_volume']) ? 50 : $modSettings['SCM_volume'];

	// Get ready to show the settings to the user:
	$context['post_url'] = $scripturl . '?action=admin;area=scm_media_player;sa=settings;save';
	$context['settings_title'] = $txt['SCM_title'];
	$context['sub_template'] = 'show_settings';
	prepareDBSettingContext($config_vars);
}

/*******************************************************************************/
// Function handling changing the CSS skin used by the mod:
/*******************************************************************************/
function SCMP_Skins()
{
	global $txt, $scripturl, $context, $settings, $modSettings;

	$styles = array(
		'aquaBlue', 'aquaGreen', 'aquaOrange', 'aquaPink', 'aquaPurple', 'aquaPurpleReal',
		'black', 'blue', 'cyber', 'green', 'orange', 'pink',  'purple', 'tunes',
		'scmBlue', 'scmGreen', 'scmOrange', 'scmPurple', 'scmRed', 'simpleBlack',
		'simpleBlue', 'simpleGreen', 'simpleOrange', 'simplePurple', 'simpleRed',
	);
	$list = array();
	foreach ($styles as $style)
		$list[$style] = $style;
	$context['SCM_styles'] = $list;

	// Use the show_settings template with our callback function:
	$config_vars = array(
		array('callback', 'SCM_style'),
	);

	// Saving?
	if (isset($_GET['save']))
	{
		checkSession();

		// Get the new style for the player:
		$_POST['SCM_custom_url'] = !empty($_POST['SCM_custom_url']) ? $_POST['SCM_custom_url'] : '';
		$config_vars[] = array('text', 'SCM_custom_url');
		$_POST['SCM_style'] = !empty($_POST['SCM_style']) ? $_POST['SCM_style'] : 'aquaBlue';
		$_POST['SCM_style'] = ($_POST['SCM_style'] == '_custom_' && !empty($_POST['SCM_custom_url']) ? '_custom' : (isset($list[$_POST['SCM_style']]) ? $_POST['SCM_style'] : 'aquaBlue'));
		$config_vars[] = array('text', 'SCM_style');
		$_POST['SCM_last_update'] = time();
		$config_vars[] = array('int', 'SCM_last_update');

		// Save the settings for this mod:
		saveDBSettings($config_vars);
		redirectexit('action=admin;area=scm_media_player;sa=skins');
	}

	// Get ready to show the settings to the user:
	$context['post_url'] = $scripturl . '?action=admin;area=scm_media_player;sa=skins;save';
	$context['settings_title'] = $txt['SCM_style_title'];
	$context['sub_template'] = 'show_settings';
	prepareDBSettingContext($config_vars);
}

/*******************************************************************************/
// Function handling listing all stored playlists:
/*******************************************************************************/
function SCMP_Lists()
{
	global $modSettings, $context, $txt, $scripturl;

	// Saving?
	if (isset($_GET['save']))
	{
		checkSession();

		// Record which playlist is the forum default:
		$_POST['SCM_last_update'] = time();
		$config_vars[] = array('int', 'SCM_last_update');
		$_POST['SCM_selected_playlist'] = !empty($_POST['SCM_selected_playlist']) ? $_POST['SCM_selected_playlist'] : 0;
		$config_vars[] = array('int', 'SCM_selected_playlist');

		// Save the settings for this mod:
		saveDBSettings($config_vars);
		redirectexit('action=admin;area=scm_media_player;sa=playlists');
	}

	// Define the template stuff we need:
	loadLanguage('ManageBoards');
	$context['page_title' ] = $txt['SCM_playlists'];
	$context['sub_template'] = 'SCMP_playlist';
	$context['post_url'] = $scripturl . '?action=admin;area=scm_media_player;sa=' . $_REQUEST['sa'];
	$modSettings['SCM_selected_playlist'] = empty($modSettings['SCM_selected_playlist']) ? 0 : $modSettings['SCM_selected_playlist'];
}

/*******************************************************************************/
// Function handling editing the playlist:
/*******************************************************************************/
function SCMP_Edit()
{
	global $txt, $scripturl, $context, $settings, $modSettings;

	if ($_REQUEST['sa'] == 'new')
	{
		// Create a new playlist entry:
		$list = max(array_keys($context['SCMP_playlists'])) + 1;
		$context['SCMP_playlists'][$list] = array(
			'id' => $list,
			'name' => sprintf($txt['SCMP_playlist_which'], $list),
			'songs' => array()
		);
	}
	else
	{
		// Attempting to edit a playlist that doesn't exist?
		$list = (int) (!empty($_GET['list']) ? $_GET['list'] : false);
		if (!isset($context['SCMP_playlists'][$list]) && !isset($_GET['save']))
			fatal_lang_error('SCM_playlist_nonexistant', false);
	}
		

	// Saving?
	if (isset($_GET['save']))
	{
		checkSession();

		// Get the new playlist ready to store in the forum settings:
		$playlist = 'SCM_playlist' . (!empty($list) ? '_' . ((int) $list) : '');
		$songs = array(
			'__NAME__' => $_POST['SCM_playlists_name'],
		);
		foreach ($_POST['SCM_title'] as $id => $title)
		{
			if (empty($_POST['SCM_url'][$id]) || empty($_POST['SCM_title'][$id]))
				unset($_POST['SCM_url'][$id], $_POST['SCM_title'][$id]);
			else
			{
				$data = $_POST['SCM_url'][$id];
				$songs[$title] = (strpos($data, 'http://') === false && strpos($data, 'https://') === false && strpos($data, '//') === false ? 'http://' : '') . $data;
			}
		}
		$serialize = function_exists('safe_serialize') ? 'safe_serialize' : 'serialize';
		$_POST[$playlist] = empty($songs) ? false : $serialize($songs);
		$config_vars[] = array('text', $playlist);
		$_POST['SCM_last_update'] = time();
		$config_vars[] = array('int', 'SCM_last_update');

		// Update the cached playlists with the new information:
		$context['SCMP_playlists'][$list] = array(
			'id' => $list,
			'name' => $songs['__NAME__'],
		);
		unset($songs['__NAME__']);
		$context['SCMP_playlists'][$list]['songs'] = $songs;
		cache_put_data('SCMP_playlists', $context['SCMP_playlists'], 86400);

		// Save the settings for this mod before redirecting to playlists:
		saveDBSettings($config_vars);
		redirectexit('action=admin;area=scm_media_player;sa=playlists');
	}

	// Use the show_settings template with our callback function:
	$config_vars = array(
		array('text', 'SCM_playlists_name', 'size' => 40),
		'',
		array('callback', 'SCM_songs'),
	);

	// Get the playlist we are editing:
	$context['SCMP_songlist'] = &$context['SCMP_playlists'][$list];
	$modSettings['SCM_playlists_name'] = $context['SCMP_songlist']['name'];

	// Get ready to show the settings to the user:
	$context['post_url'] = $scripturl . '?action=admin;area=scm_media_player;sa=' . $_REQUEST['sa'] . ($_REQUEST['sa'] == 'edit' ? ';list=' . $list : '') . ';save';
	$context['settings_title'] = $txt['SCM_playlist'] . ': ' . $context['SCMP_songlist']['name'];
	$context['sub_template'] = 'show_settings';
	prepareDBSettingContext($config_vars);
}

/*******************************************************************************/
// Function handling delete a playlist:
/*******************************************************************************/
function SCMP_Remove()
{
	global $context, $modSettings, $smcFunc;

	// Attempting to access a playlist that doesn't exist or can't be deleted?
	$list = (int) (!empty($_GET['list']) ? $_GET['list'] : false);
	if ($list === false || $list === 0)
		fatal_lang_error('SCM_cannot_remove_default', false);
	if (!isset($context['SCMP_playlists'][$list]))
		fatal_lang_error('SCM_playlist_nonexistant', false);

	// We need to delete the setting from the database:
	unset($modSettings['SCM_playlist_' . $list]);
	$smcFunc['db_query']('', '
		DELETE FROM {db_prefix}settings
		WHERE variable = {string:variable}',
		array(
			'variable' => 'SCM_playlist_' . $list,
		)
	);

	// Save the revised playlist array in the SMF cache:
	unset($context['SCMP_playlists'][$list]);
	cache_put_data('SCMP_playlists', $context['SCMP_playlists'], 86400);

	// If this playlist is selected, change back to default playlist:
	if (!empty($modSettings['SCM_selected_playlist']) && $modSettings['SCM_selected_playlist'] == $list)
	{
		$_POST['SCM_selected_playlist'] = 0;
		$config_vars[] = array('int', 'SCM_selected_playlist');
		$_POST['SCM_last_update'] = time();
		$config_vars[] = array('int', 'SCM_last_update');
	}

	// Record changes to the database and/or cache:
	$_POST['SCM_last_update'] = time();
	$config_vars[] = array('int', 'SCM_last_update');
	saveDBSettings($config_vars);
	redirectexit('action=admin;area=scm_media_player;sa=playlists');
}

?>